File: plover_vm/memory/hid.py
"""HID input state for Mailbox keyboard/mouse queues."""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

CMD_HID_POLL = 0x40
CMD_HID_KEY_READ = 0x41
CMD_HID_MOUSE_READ = 0x42
CMD_HID_INJECT = 0x43

INJECT_KEY = 0
INJECT_MOUSE = 1

KEY_MAX = 64
MOUSE_MAX = 32


@dataclass
class MouseEvent:
    buttons: int = 0
    dx: int = 0
    dy: int = 0


@dataclass
class HidState:
    """Keyboard ASCII FIFO + mouse event queue."""

    key_queue: deque[int] = field(default_factory=deque)
    mouse_queue: deque[MouseEvent] = field(default_factory=deque)
    last_key: int = 0
    last_mouse: MouseEvent = field(default_factory=MouseEvent)

    @property
    def key_pending(self) -> bool:
        return len(self.key_queue) > 0

    @property
    def mouse_pending(self) -> bool:
        return len(self.mouse_queue) > 0

    def enqueue_key(self, ch: int) -> None:
        if len(self.key_queue) >= KEY_MAX:
            self.key_queue.popleft()
        self.key_queue.append(ch & 0xFF)

    def enqueue_mouse(self, buttons: int, dx: int, dy: int) -> None:
        if len(self.mouse_queue) >= MOUSE_MAX:
            self.mouse_queue.popleft()
        self.mouse_queue.append(
            MouseEvent(buttons=buttons & 0x07, dx=_sign8(dx), dy=_sign8(dy))
        )

    def poll(self, buffer: bytearray) -> bool:
        buffer[0] = min(len(self.key_queue), 255)
        buffer[1] = min(len(self.mouse_queue), 255)
        return True

    def read_key(self, buffer: bytearray) -> bool:
        if self.key_queue:
            self.last_key = self.key_queue.popleft()
        else:
            self.last_key = 0
        buffer[0] = self.last_key & 0xFF
        return True

    def read_mouse(self, buffer: bytearray) -> bool:
        if self.mouse_queue:
            ev = self.mouse_queue.popleft()
            self.last_mouse = ev
            buffer[0] = ev.buttons & 0xFF
            buffer[1] = ev.dx & 0xFF
            buffer[2] = ev.dy & 0xFF
        else:
            self.last_mouse = MouseEvent()
            buffer[0] = 0
            buffer[1] = 0
            buffer[2] = 0
        return True

    def inject(self, buffer: bytearray | bytes) -> bool:
        if len(buffer) < 2:
            return False
        typ = buffer[0] & 0xFF
        if typ == INJECT_KEY:
            self.enqueue_key(buffer[1])
            return True
        if typ == INJECT_MOUSE:
            if len(buffer) < 4:
                return False
            self.enqueue_mouse(buffer[1], buffer[2], buffer[3])
            return True
        return False

    def dispatch(self, cmd: int, buffer: bytearray) -> bool:
        """Return True if command accepted, False if silent drop."""
        if cmd == CMD_HID_POLL:
            return self.poll(buffer)
        if cmd == CMD_HID_KEY_READ:
            return self.read_key(buffer)
        if cmd == CMD_HID_MOUSE_READ:
            return self.read_mouse(buffer)
        if cmd == CMD_HID_INJECT:
            return self.inject(buffer)
        return False


def _sign8(v: int) -> int:
    x = v & 0xFF
    return x if x < 128 else x - 256

File: crates/plover_copro/src/hid.rs
use std::collections::VecDeque;

pub const CMD_HID_POLL: u8 = 0x40;
pub const CMD_HID_KEY_READ: u8 = 0x41;
pub const CMD_HID_MOUSE_READ: u8 = 0x42;
pub const CMD_HID_INJECT: u8 = 0x43;

pub const INJECT_KEY: u8 = 0;
pub const INJECT_MOUSE: u8 = 1;

pub const KEY_MAX: usize = 64;
pub const MOUSE_MAX: usize = 32;

#[derive(Clone, Copy, Default, Debug, PartialEq, Eq)]
pub struct MouseEvent {
    pub buttons: u8,
    pub dx: i8,
    pub dy: i8,
}

#[derive(Clone, Debug, Default)]
pub struct HidState {
    key_queue: VecDeque<u8>,
    mouse_queue: VecDeque<MouseEvent>,
    pub last_key: u8,
    pub last_mouse: MouseEvent,
}

impl HidState {
    pub fn key_pending(&self) -> bool {
        !self.key_queue.is_empty()
    }

    pub fn mouse_pending(&self) -> bool {
        !self.mouse_queue.is_empty()
    }

    pub fn key_queue_len(&self) -> usize {
        self.key_queue.len()
    }

    pub fn enqueue_key(&mut self, ch: u8) {
        if self.key_queue.len() >= KEY_MAX {
            self.key_queue.pop_front();
        }
        self.key_queue.push_back(ch);
    }

    pub fn enqueue_mouse(&mut self, buttons: u8, dx: i8, dy: i8) {
        if self.mouse_queue.len() >= MOUSE_MAX {
            self.mouse_queue.pop_front();
        }
        self.mouse_queue.push_back(MouseEvent {
            buttons: buttons & 0x07,
            dx,
            dy,
        });
    }

    fn poll(&mut self, buffer: &mut [u8; 248]) -> bool {
        buffer[0] = (self.key_queue.len().min(255)) as u8;
        buffer[1] = (self.mouse_queue.len().min(255)) as u8;
        true
    }

    fn read_key(&mut self, buffer: &mut [u8; 248]) -> bool {
        self.last_key = self.key_queue.pop_front().unwrap_or(0);
        buffer[0] = self.last_key;
        true
    }

    fn read_mouse(&mut self, buffer: &mut [u8; 248]) -> bool {
        if let Some(ev) = self.mouse_queue.pop_front() {
            self.last_mouse = ev;
            buffer[0] = ev.buttons;
            buffer[1] = ev.dx as u8;
            buffer[2] = ev.dy as u8;
        } else {
            self.last_mouse = MouseEvent::default();
            buffer[0] = 0;
            buffer[1] = 0;
            buffer[2] = 0;
        }
        true
    }

    fn inject(&mut self, buffer: &[u8]) -> bool {
        if buffer.len() < 2 {
            return false;
        }
        match buffer[0] {
            INJECT_KEY => {
                self.enqueue_key(buffer[1]);
                true
            }
            INJECT_MOUSE if buffer.len() >= 4 => {
                self.enqueue_mouse(
                    buffer[1],
                    sign8(buffer[2]),
                    sign8(buffer[3]),
                );
                true
            }
            _ => false,
        }
    }

    pub fn dispatch(&mut self, cmd: u8, buffer: &mut [u8; 248], payload_len: usize) -> bool {
        let payload = &buffer[..payload_len.min(buffer.len())];
        match cmd {
            CMD_HID_POLL => self.poll(buffer),
            CMD_HID_KEY_READ => self.read_key(buffer),
            CMD_HID_MOUSE_READ => self.read_mouse(buffer),
            CMD_HID_INJECT => self.inject(payload),
            _ => false,
        }
    }
}

fn sign8(v: u8) -> i8 {
    let x = v as i16;
    if x < 128 {
        x as i8
    } else {
        (x - 256) as i8
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hid_inject_read_key() {
        let mut hid = HidState::default();
        let mut buf = [0u8; 248];
        buf[0] = INJECT_KEY;
        buf[1] = b'Z';
        assert!(hid.dispatch(CMD_HID_INJECT, &mut buf, 2));
        assert!(hid.key_pending());
        hid.dispatch(CMD_HID_POLL, &mut buf, 248);
        assert_eq!(buf[0], 1);
        hid.dispatch(CMD_HID_KEY_READ, &mut buf, 248);
        assert_eq!(buf[0], b'Z');
        assert!(!hid.key_pending());
    }

    #[test]
    fn hid_inject_read_mouse() {
        let mut hid = HidState::default();
        let mut buf = [0u8; 248];
        buf[..4].copy_from_slice(&[INJECT_MOUSE, 0x01, 0xFE, 0x05]);
        assert!(hid.dispatch(CMD_HID_INJECT, &mut buf, 4));
        assert!(hid.mouse_pending());
        hid.dispatch(CMD_HID_MOUSE_READ, &mut buf, 248);
        assert_eq!(buf[0], 0x01);
        assert_eq!(buf[1], 0xFE);
        assert_eq!(buf[2], 0x05);
    }

    #[test]
    fn hid_overflow_drop_oldest() {
        let mut hid = HidState::default();
        for i in 0..(KEY_MAX + 5) {
            hid.enqueue_key(0x30 + (i % 10) as u8);
        }
        assert_eq!(hid.key_queue_len(), KEY_MAX);
        let mut buf = [0u8; 248];
        hid.dispatch(CMD_HID_KEY_READ, &mut buf, 248);
        assert_eq!(buf[0], 0x30 + 5);
    }

    #[test]
    fn hid_invalid_inject() {
        let mut hid = HidState::default();
        let mut buf = [0u8; 248];
        buf[..2].copy_from_slice(&[99, 0]);
        assert!(!hid.dispatch(CMD_HID_INJECT, &mut buf, 2));
        buf[..2].copy_from_slice(&[INJECT_MOUSE, 0]);
        assert!(!hid.dispatch(CMD_HID_INJECT, &mut buf, 2));
    }
}
